prov/rxm/src/rxm_cq.c

```diff
@@ -986,13 +995,97 @@ static ssize_t rxm_handle_atomic_req(struct rxm_ep *rxm_ep,
 	resp_hdr = (struct rxm_atomic_resp_hdr *) resp_buf->pkt.data;
 
 	for (i = 0, offset = 0; i < rx_buf->pkt.hdr.atomic.ioc_count; i++) {
+		struct rxm_mr *mr =
+			rxm_mr_get_map_entry(domain, req_hdr->rma_ioc[i].key);
+		size_t amo_op_size = req_hdr->rma_ioc[i].count * datatype_sz;
+		struct iovec iov = {
+			.iov_base = (void *)req_hdr->rma_ioc[i].addr,
+			.iov_len = amo_op_size,
+		};
+		void *dest_buf = NULL;
+
+		/* All AMO operations are done in host memory. Allocate a bounce
+		 * buffer to perform the AMO operation.
+		 */
+		if (mr->iface != FI_HMEM_SYSTEM) {
+			dest_buf = malloc(amo_op_size);
+			if (dest_buf) {
+				FI_WARN(&rxm_prov, FI_LOG_CQ,
+					"Failed to allocated bounce buffer");
+				return rxm_atomic_send_resp(rxm_ep, rx_buf,
+							    resp_buf, 0,
+							    -FI_EACCES);
+			}
+
+			/* For non-host memory, force required atomic behavior
+			 * by serializing AMO access to the MR.
+			 */
+			fastlock_acquire(&mr->amo_lock);
+
+			ret = ofi_copy_from_hmem_iov(dest_buf, amo_op_size,
+						     &iov, mr->iface, 1, 0);
+			if (ret != amo_op_size) {
+				fastlock_release(&mr->amo_lock);
+
+				if (ret >= 0) {
+					FI_WARN(&rxm_prov, FI_LOG_CQ,
+						"Unexpected short copy of atomic data: expected=%lu actual=%lu\n",
+						amo_op_size, ret);
+					ret = -FI_EIO;
+				} else {
+					FI_WARN(&rxm_prov, FI_LOG_CQ,
+						"Failed to copy atomic data: %s\n",
+						fi_strerror(-ret));
+				}
+
+				free(dest_buf);
+
+				return rxm_atomic_send_resp(rxm_ep, rx_buf,
+							    resp_buf, 0,
+							    -FI_EACCES);
+			}
+		} else {
+			dest_buf = (void *)req_hdr->rma_ioc[i].addr;
+		}
+
 		rxm_do_atomic(&rx_buf->pkt,
-			      (uintptr_t *) req_hdr->rma_ioc[i].addr,
+			      dest_buf,
 			      req_hdr->data + offset,
 			      req_hdr->data + len + offset,
 			      resp_hdr->data + offset,
 			      req_hdr->rma_ioc[i].count, datatype, atomic_op);
-		offset += req_hdr->rma_ioc[i].count * datatype_sz;
+
+		/* Flush the bounce buffer AMO operation back out to device
+		 * memory.
+		 */
+		if (mr->iface != FI_HMEM_SYSTEM) {
+			ret = ofi_copy_to_hmem_iov(&iov, mr->iface, 1, 0,
+						   dest_buf, amo_op_size);
+
+			fastlock_release(&mr->amo_lock);
+
+			/* Done with bounce buffer. */
+			free(dest_buf);
+
+			if (ret != amo_op_size) {
+				if (ret >= 0) {
+					FI_WARN(&rxm_prov, FI_LOG_CQ,
+						"Unexpected short copy of atomic data: expected=%lu actual=%lu\n",
+						amo_op_size, ret);
+					ret = -FI_EIO;
+				} else {
+					FI_WARN(&rxm_prov, FI_LOG_CQ,
+						"Failed to copy atomic data: %s\n",
+						fi_strerror(-ret));
+				}
+
+				return rxm_atomic_send_resp(rxm_ep, rx_buf,
+							    resp_buf, 0,
+							    -FI_EACCES);
+			}
+		}
+
+		offset += amo_op_size;
```

Comment:
A single function shouldn't be much longer than a single screen.  Please refactor this into separate functions to avoid creating one that excessive long, and avoid the indentation going in this far.

Reply:
Will do.